miden-crypto/src/merkle/sparse_path.rs

```diff
@@ -0,0 +1,813 @@
+        let node = self.path.at_depth(this_depth).unwrap();
+        Some(node)
+    }
+
+    // SparseMerkleIter always knows its exact size.
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let remaining = ExactSizeIterator::len(self);
+        (remaining, Some(remaining))
+    }
+}
+
+impl ExactSizeIterator for SparseMerklePathIter<'_> {
+    fn len(&self) -> usize {
+        self.next_depth as usize
+    }
+}
+
+impl FusedIterator for SparseMerklePathIter<'_> {}
+
+// TODO: impl DoubleEndedIterator.
+
+/// Owning iterator for [SparseMerklePath].
+pub struct IntoIter {
+    /// The "inner" value we're iterating over.
+    path: SparseMerklePath,
+
+    /// The depth a `next()` call will get. `next_depth == 0` indicates that the iterator has been
+    /// exhausted.
+    next_depth: u8,
+}
+
+impl IntoIterator for SparseMerklePath {
+    type IntoIter = IntoIter;
+    type Item = <Self::IntoIter as Iterator>::Item;
+
+    fn into_iter(self) -> IntoIter {
+        let tree_depth = self.depth();
+        IntoIter { path: self, next_depth: tree_depth }
+    }
+}
+
+impl Iterator for IntoIter {
+    type Item = RpoDigest;
+
+    fn next(&mut self) -> Option<RpoDigest> {
+        let this_depth = self.next_depth;
+        // Paths don't include the root, so if `this_depth` is 0 then we keep returning `None`.
+        let this_depth = NonZero::new(this_depth)?;
+        self.next_depth = this_depth.get() - 1;
+
+        // `this_depth` is only ever decreasing, so it can't ever exceed `self.path.depth()`.
+        let node = self.path.at_depth(this_depth).unwrap();
+        Some(node)
+    }
+
+    // IntoIter always knows its exact size.
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let remaining = ExactSizeIterator::len(self);
+        (remaining, Some(remaining))
+    }
+}
+
+impl ExactSizeIterator for IntoIter {
+    fn len(&self) -> usize {
+        self.next_depth as usize
+    }
+}
+
+impl FusedIterator for IntoIter {}
+
+// TODO: impl DoubleEndedIterator.
+
+// COMPARISONS
+// ================================================================================================
+impl PartialEq<MerklePath> for SparseMerklePath {
+    fn eq(&self, rhs: &MerklePath) -> bool {
+        if self.depth() != rhs.depth() {
+            return false;
+        }
+
+        for (node, &rhs_node) in iter::zip(self, rhs.iter()) {
+            if node != rhs_node {
+                return false;
+            }
+        }
+
+        true
+    }
+}
+
+impl PartialEq<SparseMerklePath> for MerklePath {
+    fn eq(&self, rhs: &SparseMerklePath) -> bool {
+        rhs == self
+    }
+}
+
+// SPARSE VALUE PATH
+// ================================================================================================
+/// A container for a [crate::Word] value and its [SparseMerklePath] opening.
+#[derive(Clone, Debug, Default, PartialEq, Eq)]
+pub struct SparseValuePath {
```

Comment:
I wonder if we should call this `SimpleSmtProof` instead (if this ends up replacing `ValuePath` completely), to be analogous to `SmtProof`.

Reply:
Can make the change now, or we can consider it part of the https://github.com/0xPolygonMiden/crypto/issues/404 ? What would you prefer?